File: python_files/raylib/inventory.py
import pyray as p
from dataclasses import dataclass

SCREEN_WIDTH: int = 800
SCREEN_HEIGHT: int = 600
SCREEN_TITLE: str = "Inventory"
SCREEN_BACKGROUND: p.Color = p.RAYWHITE

ROWS: int = 5
COLS: int = 5
SLOT_SIZE: int = 80
SLOT_GAP: int = 3

NUM_OF_ITEMS: int = 3


@dataclass
class Item:
    id: int
    rect: p.Rectangle
    color: p.Color
    offset: p.Vector2
    dragging: bool
    quantity: int


def main() -> None:
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE)

    pos: p.Vector2 = p.Vector2(10, 10)

    slots: list[p.Rectangle] = []
    for i in range(ROWS):
        for j in range(COLS):
            x: int = j * (SLOT_SIZE + SLOT_GAP) + pos.x
            y: int = i * (SLOT_SIZE + SLOT_GAP) + pos.y

            slots.append(p.Rectangle(x, y, SLOT_SIZE, SLOT_SIZE))

    inventory: list[Item] = []
    inventory_active: bool = True

    items: list[Item] = []
    for i in range(NUM_OF_ITEMS):
        item: Item = Item(
            id=(i + 1),
            rect=p.Rectangle(600, (i + 1) * 80, 50, 50),
            color=p.Color(
                p.get_random_value(0, 255),
                p.get_random_value(0, 255),
                p.get_random_value(0, 255),
                255,
            ),
            dragging=False,
            quantity=1,
            offset=p.Vector2(0, 0),
        )

        items.append(item)

        last_current_time: float = 0.0
        print_interval: float = 1.0

    while not p.window_should_close():
        # activate/deactivate inventory
        if p.is_key_pressed(p.KeyboardKey.KEY_I):
            inventory_active = not inventory_active

        mouse_pos: p.Vector2 = p.get_mouse_position()

        # activate item dragging
        for item in items:
            if p.is_mouse_button_pressed(p.MouseButton.MOUSE_BUTTON_LEFT):
                if (
                    p.check_collision_point_rec(mouse_pos, item.rect)
                    and not item.dragging
                ):
                    item.dragging = True
                    item.offset.x = mouse_pos.x - item.rect.x
                    item.offset.y = mouse_pos.y - item.rect.y

        for item in inventory[:]:
            if p.is_mouse_button_pressed(p.MouseButton.MOUSE_BUTTON_LEFT):
                if p.check_collision_point_rec(mouse_pos, item.rect) and not item.dragging:
                    item.dragging = True
                    item.offset.x = mouse_pos.x - item.rect.x
                    item.offset.y = mouse_pos.y - item.rect.y

                    inventory.remove(item)
                    items.append(item)

        # handle dragging
        for item in items[:]:
            if item.dragging:
                item.rect.x = mouse_pos.x - item.offset.x
                item.rect.y = mouse_pos.y - item.offset.y

                if p.is_mouse_button_released(p.MouseButton.MOUSE_BUTTON_LEFT):
                    item.dragging = False

                    # snap item to slot
                    for slot in slots:
                        if p.check_collision_recs(slot, item.rect):
                            item.rect.x = slot.x + (slot.width - item.rect.width) / 2
                            item.rect.y = slot.y + (slot.height - item.rect.height) / 2

                            inventory.append(item)
                            if item in items:
                                items.remove(item)

        # debug prints
        current_time: float = p.get_time()
        if current_time - last_current_time > print_interval:
            last_current_time = current_time
            print()
            print(f"Time: {last_current_time}")
            print()
            print("========== Items ===============")
            for item in items:
                print(f"id: {item.id}; x: {item.rect.x}; y: {item.rect.y}")
            print()
            print("========== Inventory ===============")
            for item in inventory:
                print(f"id: {item.id}; x: {item.rect.x}; y: {item.rect.y}")
            print()

        p.begin_drawing()
        p.clear_background(SCREEN_BACKGROUND)

        if inventory_active:
            # draw slots
            for slot in slots:
                slot_color: p.Color = (
                    p.RED if p.check_collision_point_rec(mouse_pos, slot) else p.BLACK
                )

                p.draw_rectangle_lines_ex(slot, 5, slot_color)

            # draw inventory items
            for item in inventory:
                p.draw_rectangle_rec(item.rect, item.color)

        # draw items
        for item in items:
            p.draw_rectangle_rec(item.rect, item.color)

        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()

File: c_files/raylib/inventory.c
#include <raylib.h>
#include <stdio.h>

#define ROWS 5
#define COLS 5
#define SLOT_SIZE 80
#define SLOT_GAP 5

#define iITEMS_SIZE 5

typedef struct Item
{
  int id;
  Rectangle rect;
  Color color;
  Vector2 offset;
  bool dragging;

} Item;

typedef struct Items
{
  Item list[iITEMS_SIZE];
  int used;

} Items;

typedef struct Inventory
{
  Item list[ROWS * COLS];
  int used;
  bool active;

} Inventory;

int main(void)
{
  const int SCREEN_WIDTH = 800;
  const int SCREEN_HEIGHT = 600;
  const char SCREEN_TITLE[] = "Inventory";
  const Color SCREEN_BACKGROUND = RAYWHITE;

  InitWindow(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE);

  Vector2 pos = (Vector2){10, 10};

  Rectangle slots[ROWS * COLS];
  for (int i = 0; i < ROWS; i++)
  {
    for (int j = 0; j < COLS; j++)
    {
      int x = j * (SLOT_SIZE + SLOT_GAP) + pos.x;
      int y = i * (SLOT_SIZE + SLOT_GAP) + pos.y;

      slots[i * COLS + j] = (Rectangle){x, y, SLOT_SIZE, SLOT_SIZE};
    }
  }

  Inventory inventory;
  inventory.active = true;
  inventory.used = 0;

  Items items;
  items.used = 0;

  for (int i = 0; i < iITEMS_SIZE; i++)
  {
    Item item;
    item.id = i;
    item.rect = (Rectangle){600, (i + 1) * 70, 50, 50};
    item.color = (Color){
        GetRandomValue(0, 255),
        GetRandomValue(0, 255),
        GetRandomValue(0, 255),
        255,
    };
    item.offset = (Vector2){0, 0};
    item.dragging = false;

    items.list[i] = item;
    items.used++;
  }

  float last_current_time = 0.0f;
  float print_interval = 1.0f;

  while (!WindowShouldClose())
  {
    // activate-deactivate inventory
    if (IsKeyPressed(KEY_I))
    {
      inventory.active = !inventory.active;
    }

    Vector2 mouse_pos = GetMousePosition();

    // inventory list
    for (int i = 0; i < inventory.used; i++)
    {
      if (IsMouseButtonPressed(MOUSE_LEFT_BUTTON))
      {
        if (CheckCollisionPointRec(mouse_pos, inventory.list[i].rect) && !inventory.list[i].dragging)
        {
          inventory.list[i].dragging = true;
          inventory.list[i].offset.x = mouse_pos.x - inventory.list[i].rect.x;
          inventory.list[i].offset.y = mouse_pos.y - inventory.list[i].rect.y;

          // add item to inventory
          if (items.used < iITEMS_SIZE)
          {
            items.list[items.used] = inventory.list[i];
            items.used++;
          }

          // remove item from inventory
          if (inventory.used > 0)
          {
            for (int j = i; j < inventory.used; j++)
            {
              inventory.list[j] = inventory.list[j + 1];
            }
            inventory.used--;
          }
        }
      }
    }

    // handle dragging
    for (int i = 0; i < items.used; i++)
    {
      if (IsMouseButtonPressed(MOUSE_LEFT_BUTTON))
      {
        if (CheckCollisionPointRec(mouse_pos, items.list[i].rect) && !items.list[i].dragging)
        {
          items.list[i].dragging = true;
          items.list[i].offset.x = mouse_pos.x - items.list[i].rect.x;
          items.list[i].offset.y = mouse_pos.y - items.list[i].rect.y;
        }
      }

      if (items.list[i].dragging)
      {
        items.list[i].rect.x = mouse_pos.x - items.list[i].offset.x;
        items.list[i].rect.y = mouse_pos.y - items.list[i].offset.y;

        if (IsMouseButtonReleased(MOUSE_LEFT_BUTTON))
        {
          items.list[i].dragging = false;

          for (int j = 0; j < ROWS * COLS; j++)
          {
            if (CheckCollisionRecs(items.list[i].rect, slots[j]))
            {
              // snap item to slot
              items.list[i].rect.x = slots[j].x + (slots[j].width - items.list[i].rect.width) / 2;
              items.list[i].rect.y = slots[j].y + (slots[j].height - items.list[i].rect.height) / 2;

              // add item to inventory
              if (inventory.used < ROWS * COLS)
              {
                inventory.list[inventory.used] = items.list[i];
                inventory.used++;
              }

              // remove item from items list
              if (items.used > 0)
              {
                for (int k = i; k < items.used; k++)
                {
                  items.list[k] = items.list[k + 1];
                }
                items.used--;
              }

              break;
            }
          }
        }
      }
    }

    // debug print statements
    float current_time = GetTime();
    if (current_time - last_current_time > print_interval)
    {
      last_current_time = current_time;

      printf("\nTime: %f\n", current_time);
      printf("\nItems: \n");
      for (int i = 0; i < items.used; i++)
      {
        printf("Id: %d\nX: %.2f\nY: %.2f\n", items.list[i].id,
               items.list[i].rect.x,
               items.list[i].rect.y);
      }
      printf("\nInventory: \n");
      for (int i = 0; i < inventory.used; i++)
      {
        printf("Id: %d\nX: %.2f\nY: %.2f\n", inventory.list[i].id,
               inventory.list[i].rect.x,
               inventory.list[i].rect.y);
      }
    }

    BeginDrawing();
    ClearBackground(SCREEN_BACKGROUND);

    for (int i = 0; i < items.used; i++)
    {
      DrawRectangleRec(items.list[i].rect, items.list[i].color);
    }

    if (inventory.active)
    {
      // draw slots
      for (int i = 0; i < ROWS * COLS; i++)
      {
        Color slot_color = CheckCollisionPointRec(mouse_pos, slots[i]) ? RED : BLACK;

        DrawRectangleLinesEx(slots[i], 5, slot_color);
      }

      // draw inventory items
      for (int i = 0; i < inventory.used; i++)
      {
        DrawRectangleRec(inventory.list[i].rect, inventory.list[i].color);
      }
    }

    EndDrawing();
  }

  CloseWindow();

  return 0;
}